src/impl/sharding.cc

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*!
+ * \file src/impl/sharding.cc
+ * \brief Description of RAF sharding specifications
+ */
+#include <tvm/runtime/data_type.h>
+#include "raf/ir.h"
+#include "raf/op.h"
+#include "raf/op_utils.h"
+#include "raf/type.h"
+#include "raf/registry.h"
+#include "raf/sharding.h"
+#include "raf/communicator.h"
+#include "../op/ty/utils.h"
+#include "../op/schema/ufunc.h"
+#include "../op/schema/sharding.h"
+#include "../op/dialect/tvm/tvm_utils.h"
+#include "../op/dialect/tvm/tvm_attrs.h"
+#include <string>
+
+namespace raf {
+namespace sharding {
+
+using namespace raf::ir;
+using namespace raf::op;
+using namespace raf::op::schema;
+using namespace raf::value;
+using namespace raf::distributed;
+using namespace raf::distributed::communicator;
+
+int64_t ShardSpec::GetRankIdx(Array<Integer> ranks) {
+  for (int64_t i = 0; i < ranks.size(); ++i) {
+    if (GetGlobalCommunicator()->rank == ranks[i]->value) {
+      return i;
+    }
+  }
+  return -1;
+}
```

Comment:
It seems to me that this function tries to find the first available rank, but I don't get the reason behind it. Could you elaborate?

Reply:
For example, when `spec.ranks = [3, 2, 1, 0]` and we need to partition `array = [seg1, seg2, seg3, seg4]`, the rank 0 is responsible for maintaining `seg4` because it appears in the list as the 4th element. This function helps to identify the position where the current rank appears in the list.

The reason why we don't restrict elements in the list keep in order is it is easy for me to test. In the previous example, I can use rank 0 to read out the value of `seg4`. Another potential reason is reordering the list may get performance improvement if certain rank order + certain Op is friendly to the hardware topology.

Reply:
I see, so this function returns the index of the current rank in the list. Please add docstring, like `Returns the index of the current rank in the given rank array`, to avoid future confusion.

Reply:
I actually thinking should we remove and inline this function. Originally this function is referred multiple times, however, since the code is largely simplified and there is only one reference in the code.

Reply:
Feel free to do so. In any way you choose, however, please have sufficient comments to avoid any possible confusion.